crates/sui-types/src/temporary_store.rs

```diff
@@ -864,6 +871,181 @@ impl<S: ObjectStore> TemporaryStore<S> {
+impl<S: ObjectStore> TemporaryStore<S> {
+    /// 1. Compute tx storage gas costs and tx storage rebates, update storage_rebate field of mutated objects
+    /// 2. Deduct computation gas costs and storage costs to `gas_object_id`, credit storage rebates to `gas_object_id`.
+    // The happy path of this function follows (1) + (2) and is fairly simple. Most of the complexity is in the unhappy paths:
+    // - if execution aborted before calling this function, we have to dump all writes + re-smash gas, then charge for storage
+    // - if we run out of gas while charging for storage, we have to dump all writes + re-smash gas, then charge for storage again
+    pub fn charge_gas<T>(
+        &mut self,
+        gas_object_id: ObjectID,
+        gas_status: &mut SuiGasStatus<'_>,
+        execution_result: &mut Result<T, ExecutionError>,
+        gas: &[ObjectRef],
+    ) -> GasCostSummary {
+        // at this point, we have done *all* charging for computation,
+        // but have not yet set the storage rebate or storage gas units
+        assert!(gas_status.storage_rebate() == 0);
+        assert!(gas_status.storage_gas_units() == 0);
+
+        // println!("GAS - charge gas with result error: {}", execution_result.is_err());
+
+        // bucketize computation cost
+        if let Err(err) = gas_status.bucketize_computation() {
+            // println!("GAS - bucketize failed");
+            if execution_result.is_ok() {
+                *execution_result = Err(err);
+            }
+        }
+
+        // On error we need to dump writes, deletes, etc before charging storage gas
+        if execution_result.is_err() {
+            self.reset(gas, gas_status);
+        }
+
+        // collect and charge storage cost
+        self.collect_storage_and_rebate(gas_status, gas_object_id);
+        if let Err(err) = gas_status.charge_storage_and_rebate() {
+            // println!("GAS - Out Of Gas charging storage");
+            // out of gas
+            self.reset(gas, gas_status);
+            self.collect_storage_and_rebate(gas_status, gas_object_id);
+            // we do not charge for storage at this point because we are out of gas
+            gas_status.adjust_computation_on_out_of_gas();
+            if execution_result.is_ok() {
+                *execution_result = Err(err);
+            }
+        }
+
+        let cost_summary = gas_status.summary();
+        let gas_used = cost_summary.net_gas_usage();
+
+        // Important to fetch the gas object here instead of earlier, as it may have been reset
+        // previously in the case of error.
+        let mut gas_object = self.read_object(&gas_object_id).unwrap().clone();
+        gas::deduct_gas(&mut gas_object, gas_used);
+        trace!(gas_used, gas_obj_id =? gas_object.id(), gas_obj_ver =? gas_object.version(), "Updated gas object");
+
+        self.write_object(gas_object, WriteKind::Mutate);
+        self.gas_charged = Some((gas_object_id, cost_summary.clone()));
```

Comment:
I think we should be able to get rid of the `gas_charged` field now that you have refactored to return the summary?